<launch>
   
    <!-- 
    from bag: rosbag record /stereo_camera/left/image_raw_throttle/compressed /stereo_camera/right/image_raw_throttle/compressed /stereo_camera/left/camera_info_throttle /stereo_camera/right/camera_info_throttle /tf
    -->
    <param name="use_sim_time" type="bool" value="True"/>
        
    <node name="republish_left" type="republish" pkg="image_transport" args="compressed in:=/stereo_camera/left/image_raw_throttle raw out:=/stereo_camera/left/image_raw_throttle_relay" />
    <node name="republish_right" type="republish" pkg="image_transport" args="compressed in:=/stereo_camera/right/image_raw_throttle raw out:=/stereo_camera/right/image_raw_throttle_relay" />
  	    
  	<!-- Run the ROS package stereo_image_proc for image rectification-->
	<group ns="/stereo_camera" >
	    <node pkg="nodelet" type="nodelet" name="stereo_nodelet"  args="manager"/>
	
		<node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
			<remap from="left/image_raw"    to="left/image_raw_throttle_relay"/>
			<remap from="left/camera_info"  to="left/camera_info_throttle"/>
			<remap from="right/image_raw"   to="right/image_raw_throttle_relay"/>
			<remap from="right/camera_info" to="right/camera_info_throttle"/>
			<param name="disparity_range" value="128"/>
		</node>
		
		<node pkg="nodelet" type="nodelet" name="disparity2cloud" args="load rtabmap/point_cloud_xyz stereo_nodelet">
			<remap from="disparity/image"       to="disparity"/>
			<remap from="disparity/camera_info" to="right/camera_info_throttle"/>
			<remap from="cloud"                 to="cloudXYZ"/>
			
			<param name="voxel_size" type="double" value="0.05"/>
			<param name="decimation" type="int" value="4"/>
			<param name="max_depth" type="double" value="4"/>
		</node>
	</group>
  	 	
	<!-- Odometry: Choose between viso2_ros, fovis_ros and homemade approach (default) -->
	<!--
	<node pkg="viso2_ros" type="stereo_odometer" name="stereo_odometer" output="screen">
      <remap from="stereo" to="/stereo_camera"/>
      <remap from="image" to="image_rect"/>

      <param name="base_link_frame_id" value="base_footprint"/>
      <param name="odom_frame_id" value="odom"/>
      <param name="ref_frame_change_method" value="1"/>
      <param name="queue_size" value="30"/>
   </node>
   <node name="camInfoL" type="relay" pkg="topic_tools" args="/stereo_camera/left/camera_info_throttle /stereo_camera/left/camera_info"/>
   <node name="camInfoR" type="relay" pkg="topic_tools" args="/stereo_camera/right/camera_info_throttle /stereo_camera/right/camera_info"/>	
   -->
	<!--
	<node pkg="fovis_ros" type="fovis_stereo_odometer" name="stereo_odometer" >
	    <remap from="/stereo/left/image" to="/stereo_camera/left/image_rect" />
	    <remap from="/stereo/right/image" to="/stereo_camera/right/image_rect" />
	    <remap from="/stereo/left/camera_info" to="/stereo_camera/left/camera_info_throttle" />
	    <remap from="/stereo/right/camera_info" to="/stereo_camera/right/camera_info_throttle" />
	    <remap from="odometry" to="stereo_odometer/odometry" />
    </node>
    -->	
    
    <node pkg="rtabmap" type="stereo_odometry" name="stereo_odometry" output="screen">
      <remap from="left/image_rect"       to="/stereo_camera/left/image_rect"/>
      <remap from="right/image_rect"      to="/stereo_camera/right/image_rect"/>
      <remap from="left/camera_info"      to="/stereo_camera/left/camera_info_throttle"/>
      <remap from="right/camera_info"     to="/stereo_camera/right/camera_info_throttle"/>
      <remap from="odom" to="/stereo_odometer/odometry"/>

      <remap from="odom_local_map" to="/odom_local_map"/>
      <remap from="odom_last_frame" to="/odom_last_frame"/>

      <param name="frame_id" type="string" value="base_footprint"/>
      <param name="odom_frame_id" type="string" value="odom"/>
      <param name="approx_sync" type="bool" value="false"/>
      <param name="queue_size" type="int" value="5"/>

      <param name="Odom/Strategy" type="string" value="0"/>
      <param name="Odom/FeatureType" type="string" value="6"/>
      <param name="Odom/InlierDistance" type="string" value="0.1"/>
      <param name="Odom/MinInliers" type="string" value="10"/>
     
      <param name="Odom/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
      <param name="Odom/MaxFeatures" type="string" value="0"/>
      <param name="Odom/MaxDepth" type="string" value="10"/>
      <param name="Odom/FeaturesRatio" type="string" value="0"/>
      <param name="Odom/Iterations" type="string" value="30"/>
      <param name="Odom/RefineIterations" type="string" value="5"/>
      
      <param name="GFTT/MaxCorners" type="string" value="500"/>
      <param name="GFTT/MinDistance" type="string" value="5"/>
      
      <param name="OdomBow/LocalHistorySize" type="string" value="1000"/>
      <param name="OdomBow/NNDR" type="string" value="0.8"/>
      
      <param name="OdomFlow/WinSize" type="string" value="16"/>
      <param name="OdomFlow/MaxLevel" type="string" value="3"/>
      <param name="OdomFlow/Iterations" type="string" value="30"/>
      <param name="OdomFlow/Eps" type="string" value="0.01"/>
      
      <param name="OdomSubPix/Iterations" type="string" value="20"/>
      <param name="OdomSubPix/WinSize" type="string" value="0"/>
      <param name="OdomSubPix/Eps" type="string" value="0.02"/>
    </node>
  	
      
  <group ns="rtabmap">	
    <!-- Visual SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap" type="rtabmap" output="screen" args="--delete_db_on_start">
          <param name="frame_id" type="string" value="base_footprint"/>
          <param name="subscribe_stereo" type="bool" value="true"/>
          <param name="subscribe_depth" type="bool" value="false"/>

          <remap from="left/image_rect" to="/stereo_camera/left/image_rect_color"/>
          <remap from="right/image_rect" to="/stereo_camera/right/image_rect"/>
          <remap from="left/camera_info" to="/stereo_camera/left/camera_info_throttle"/>
          <remap from="right/camera_info" to="/stereo_camera/right/camera_info_throttle"/>

          <remap from="odom" to="/stereo_odometer/odometry"/>
          <remap from="mapData" to="mapData_notOptimized"/>

          <param name="queue_size" type="int" value="30"/>

          <!-- RTAB-Map's parameters -->
          <param name="Rtabmap/TimeThr" type="string" value="700"/>
          <param name="Rtabmap/DetectionRate" type="string" value="1"/>
          <param name="RGBD/ToroIterations" type="string" value="0"/> <!-- Disable graph optimization because we use map_optimizer node below -->

          <param name="Kp/TfIdfLikelihoodUsed" type="string" value="true"/>
          <param name="Kp/DetectorStrategy" type="string" value="0"/>
          <param name="Kp/WordsPerImage" type="string" value="200"/>
          <param name="Kp/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
          <param name="Kp/NNStrategy" type="string" value="1"/>

          <param name="Bayes/FullPredictionUpdate" type="string" value="false"/>

          <param name="SURF/HessianThreshold" type="string" value="1000"/>

          <param name="LccBow/MaxDepth" type="string" value="3"/>
          <param name="LccBow/MinInliers" type="string" value="10"/>
          <param name="LccBow/InlierDistance" type="string" value="0.02"/>

          <param name="LccReextract/Activated" type="string" value="true"/>
          <param name="LccReextract/FeatureType" type="string" value="4"/>
          <param name="LccReextract/MaxWords" type="string" value="400"/>
          <param name="LccReextract/NNDR" type="string" value="0.8"/>
          <param name="LccReextract/NNType" type="string" value="3"/>
    </node>
    
    
    <node pkg="rtabmap" type="map_optimizer" name="map_optimizer">
          <remap from="mapData" to="mapData_notOptimized"/>
          <remap from="mapData_notOptimized_optimized" to="mapData"/>
    </node>
    <node pkg="rtabmap" type="map_assembler" name="map_assembler">
          <param name="occupancy_grid" type="bool" value="true"/>
          <remap from="grid_projection_map" to="/map"/>
    </node>
        
  </group>
  
  <!-- RVIZ --> 
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap)/launch/config/azimut3_nav.rviz"/>
  
  <!-- RTABMAP GUI -->
  <!--
  <node pkg="rtabmap" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_depth" type="bool" value="false"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
      <param name="queue_size" type="int" value="30"/>
    
      <remap from="rgb/image" to="/stereo_camera/left/image_rect_color"/>
      <remap from="rgb/camera_info" to="/stereo_camera/left/camera_info_throttle"/>
      <remap from="depth/image" to="/odom_depth"/>
      
      <remap from="odom" to="/stereo_odometer/odometry"/>
    </node>
    -->
</launch>
